DEV-948 change rendering system to use SVGs instead of PNGs

torbensky opened the pull request:
**Objective**

We want to switch to SVGs sourced from the Battlesnake CDN instead of using local PNG's.

- Change the rendering system to use SVGs instead of PNGs
- Source SVGs from media.battlesnake.com
  - Snake heads: /snakes/heads/*
  - Snake tails: /snakes/tails/*

**Summary of Changes**

- Updated image library to pull SVG's of custom heads/tails from the Battlesnake media server
- Added test coverage for image library changes
- Updated GitHub actions workflow to include inkscape
- Removed local PNG's (except for default to support graceful fallback)
- Updated exporter container (add inkscape)
- Updated readme documentation
- Added additional details to documentation of `/avatar` endpoint
- Added test coverage for `/avatar` customization happy paths

torbensky committed:
add docs and tests for /avatar endpoint

torbensky committed:
Merge branch 'main' into dev-948-switch-to-s3-svgs

torbensky committed:
remove local customization assets

torbensky committed:
update test for new default asset

torbensky committed:
basic wrapper

torbensky committed:
refactoring image library to improve devx
- no panics
- logging
- testing
- cleaner APIs

torbensky committed:
use SVG content from media server

- first try to download SVG from media server
- use inkscape to rasterize
- fallback to local PNG if things fail

torbensky committed:
add SVG test coverage, fix bugs

- fix some bugs with file handling
- fixed a bug with file path
- add additional test coverage
- add some logging

torbensky committed:
fix bug with incorrect asset type used

torbensky committed:
reduce log spam

torbensky commented on render/board.go:
```diff
@@ -142,6 +142,11 @@ func (b *Board) addHazard(p *engine.Point) {
 }
 
 func getDirection(p engine.Point, nP engine.Point) snakeDirection {
+	// handle when the points haven't changed (a common case) by defaulting to up
```
this was really spammy in the logs because it is common to have these two points be the same (nothing is actually wrong)

bvanvugt replied:
Makes sense. For reasons far too ridiculous to explain... can you set the default to face right rather than up? 

It's kind of a thing in the Battlesnake community :)
<img width="381" alt="Screen Shot 2022-02-07 at 7 41 23 PM" src="https://user-images.githubusercontent.com/1531419/152913853-590e7914-a197-47a1-8863-ec8c013253eb.png">

torbensky commented on render/image.go:
```diff
@@ -1,37 +1,61 @@
 
-var assetImageCache = make(map[string]image.Image)
-var assetImageCacheLock sync.Mutex
+var boardImageCache = cache.New(6*time.Hour, time.Minute)
```
use the cache library instead of managing our own implementation of a cache

bvanvugt replied:
See note about possibly burying this inside the media client

torbensky commented on render/image.go:
```diff
@@ -57,72 +81,72 @@ func parseHexColor(x string) color.Color {
 
-	assetImage, _, err := image.Decode(f)
-	if err != nil {
-		panic(err)
```
panic == bad!

torbensky commented on render/image.go:
```diff
@@ -57,72 +81,72 @@ func parseHexColor(x string) color.Color {
+func loadImageFile(path string) (image.Image, error) {
+	f, err := os.Open(path)
 	if err != nil {
-		panic(err)
```
panic == bad

torbensky committed:
fix error message

torbensky committed:
add inkscape info to readme

torbensky committed:
add inkscape to GitHub action

torbensky committed:
fix step name copypasta

torbensky committed:
add more context to error output

torbensky committed:
install recent version of inkscape

torbensky committed:
add inkscape to container

torbensky committed:
improve build/run documentation

- docker stuff
- build to local binary instead of install

torbensky commented on .github/workflows/test.yml:
```diff
@@ -38,6 +38,11 @@ jobs:
     - uses: actions/setup-go@master
       with:
         go-version: '1.17'
+    - name: Install inkscape
```
Adds inkscape to the GitHub action so tests will pass

torbensky commented on .gitignore:
```diff
@@ -4,5 +4,7 @@
 .vscode
 .idea
 
+bin/
```
I updated the docs to suggest building a local binary to this path

bvanvugt replied:
Sounds good - in other repos we use `/bin` for scripts we want to include in the Docker build, so this may change at some point. Not a problem now though.

torbensky commented on Dockerfile:
```diff
@@ -12,7 +12,7 @@ FROM alpine:latest
 ENV APP_VERSION=$APP_VERSION
 
-RUN apk add --no-cache ca-certificates
+RUN apk add --no-cache ca-certificates inkscape
```
this adds inkscape to the exporter container (I was really happy that there was an api package for it!)

bvanvugt replied:
Hah, and an alpine package! This was much easier than I was expecting :)

torbensky commented on render/image.go:
```diff
@@ -303,15 +350,12 @@ func createBoardContext(b *Board) *gg.Context {
 
 	cacheKey := fmt.Sprintf("%d:%d", b.Width, b.Height)
-	cachedBoardImage, ok := boardImageCache[cacheKey]
+	cachedBoardImage, ok := boardImageCache.Get(cacheKey)
```
changed this to use the cache library instead of our custom cache

torbensky commented on render/image.go:
```diff
@@ -380,3 +421,72 @@ func boardYToDrawY(dc *gg.Context, y int) float64 {
+		return nil, err
+	}
+
+	key := imageCacheKey(path, w, h, rot)
```
utilize caching to prevent redundant SVG rasterizations

torbensky committed:
warn about likely test failures

torbensky committed:
go mod tidy

bvanvugt commented on inkscape/wrapper.go:
```diff
@@ -0,0 +1,76 @@
+	return err == nil
+}
+
+// SVGToPNG raserizes the SVG at the specified path to PNG format.
```
typo on "raserizes" 😄

bvanvugt commented on inkscape/wrapper.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+// SVGToPNG raserizes the SVG at the specified path to PNG format.
+func (c Client) SVGToPNG(path string, width, height int) (image.Image, error) {
```
This also came out much simpler than I was expecting, nice

bvanvugt commented on media/api.go:
```diff
@@ -5,11 +5,14 @@ import (
 	"io/ioutil"
 	"net/http"
+
+	"github.com/sirupsen/logrus"
```
This is a minor thing, but we've been importing logrus as `log` in go code as a convention.

```log "github.com/sirupsen/logrus"```

This may have gotten through in other places too.

bvanvugt commented on render/image.go:
```diff
@@ -380,3 +421,72 @@ func boardYToDrawY(dc *gg.Context, y int) float64 {
+	return err
+}
+
+func (sm svgManager) loadSVGImageAsset(name string, aType AssetType, w, h int, rot rotations) (image.Image, error) {
```
Cool cool cool. This is great, let me see if I'm following...

* SVG Manager pulls SVG sources from media and stores on disk...? for the lifetime of the container?
* It then rasterizes PNGs for width, height, rotation and caches in-mem (short term)
* loadSVGImageAsset actually returns a fully rasterized Image object (I think that tripped me up, but I get it)
* Colorization still happens at composition time, so source .PNGs are only cached unique on source/size/orientation, but not color

Am I following that correctly?

FWIW, I'm wondering if some of these abstractions should actually live inside the media layer, ie: when requesting a source image from media, I can request it in PNG or SVG format (probably need to add size and orientation to that too). Then disk mgmt, caching, etc. remains all inside media. Thoughts?

torbensky replied:
Yep, you followed it right.

Also yeah, I think `media` would make sense as you explained :+1:

bvanvugt commented on render/image_internal_test.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+func randStr(n int) string {
+	var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
```
fun fact - sourcing random strings from

```[]rune("346789BCDFGHJKMPQRSTVWXYbcdfghjkmpqrtvwxy")```

avoids accidental bad word generation :)

bvanvugt replied:
I tried to find a source for this but I can't. Just a string I've been carrying around with me for the past 10 years... 😆

torbensky replied:
TIL! (and that is a fun fact)

torbensky committed:
default right

(you can't be wrong if you're right)

torbensky committed:
change rand string so it cant be bad words

torbensky committed:
use "log" aliased logrus

Battlesnake convention is to always import as "log"

torbensky committed:
refactor media things out of render package

- move media files to media package
- move SVG/PNG loading to media package
- add test coverage
- improved the testing by mocking media server (no flake)

torbensky committed:
update docker container to use new asset location

torbensky committed:
polish (naming and comments)

bvanvugt commented on render/board.go:
```diff
@@ -142,6 +142,11 @@ func (b *Board) addHazard(p *engine.Point) {
 }
 
 func getDirection(p engine.Point, nP engine.Point) snakeDirection {
+	// handle when the points haven't changed (a common case) by defaulting to up
```
comment says 'up' still

torbensky replied:
woops!

bvanvugt commented on render/board.go:
```diff
@@ -178,7 +183,7 @@ func getDirection(p engine.Point, nP engine.Point) snakeDirection {
 
 	// default to "up" when invalid moves are passed
```
comment again, lol :)

bvanvugt commented on render/image.go:
```diff
@@ -158,26 +121,39 @@ func drawHazard(dc *gg.Context, bx int, by int) {
 
-func drawSnakeImage(filename string, fallbackFilename string, dc *gg.Context, bx int, by int, hexColor string, dir snakeDirection) {
-	var rotation int
+func drawSnakeImage(name string, st snakeImageType, dc *gg.Context, bx int, by int, hexColor string, dir snakeDirection) {
```
Nice, this came out super straight-forward.

torbensky committed:
update comment to reflect new default
